src/main/java/io/cryostat/messaging/notifications/NotificationSource.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright The Cryostat Authors
+ *
+ * The Universal Permissive License (UPL), Version 1.0
+ *
+ * Subject to the condition set forth below, permission is hereby granted to any
+ * person obtaining a copy of this software, associated documentation and/or data
+ * (collectively the "Software"), free of charge and under any and all copyright
+ * rights in the Software, and any and all patent rights owned or freely
+ * licensable by each licensor hereunder covering either (i) the unmodified
+ * Software as contributed to or provided by such licensor, or (ii) the Larger
+ * Works (as defined below), to deal in both
+ *
+ * (a) the Software, and
+ * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
+ * one is included with the Software (each a "Larger Work" to which the Software
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.messaging.notifications;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+@SuppressWarnings("unchecked")
+public class NotificationSource {
```

Comment:
Is there any way this can be refactored and merged with the `NotificationFactory`?

Reply:
What kind of merging? From the way I look at it, the `NotificationFactory` is it's own factory to make `Notification.Builder`s and the `NotificationSource` is just a source point to add/remove listeners to and notify and classes that implement `NotificationListener`, and they seem sort of separate, so I'm unsure of the best way to cleaning refactor them into one class.

Reply:
I was just thinking that if the definitions of the `NotificationSource` and `NotificationFactory` were combined, then `Notification#send()` could directly notify the listeners. It would avoid passing the Notification, cast to a `WsMessage`, into the `MessagingServer`, just for that to then check if the `WsMessage instanceof Notification` and notify the listeners if so. `MessagingServer` would retain its simple responsibility of emitting WebSocket messages to connected clients, and not also become responsible for passing events between components within the Cryostat process.

Reply:
Is what I recently pushed sort of what you were thinking?

Reply:
Sure, that looks like a good approach to achieve what I was describing before. Nice way to do it.